### ISSUE TITLE
wikipedia_en_all_maxi_2018-10.zim not downloading

### ISSUE THREAD

**Issue by Popolechien** [kiwix/apple#210]
I've had several users reporting that wikipedia_en_all_maxi_2018-10.zim fails to download (or rather: starts downloading and then stalls). They have also indicated that other files (including wikipedia_en_all_nopic_2019-12.zim) downloaded just fine, so the error might be with the handling of this one big guy.

**Comment by kelson42** [comment]
Seems a duplicate of infamous #151.

**Comment by kelson42** [comment]
@automactic What is the strategy to handle such reports? We have basically no details.

**Comment by automactic** [comment]
In general, we should close issues without detail as unable to reproduce after a certain amount of time, since we really don't know what is wrong.

This one, however, I think I have a pretty good idea about what's going on. Right now, the app does not handle the file corrupted case very well. This is why you see the file downloaded to 100% (or like 99.9%), then the progress hangs / stalls. The app moves the file from downloading to on device upon successfully opening a file. And it won't be moved if the file cannot be opened. 

What would be nice is to show user the file name and let them know you got this file, but we cannot open it. And offer to checksum if possible, so that we know if it is the download or the file itself that is problematic.

Right now, I am not gonna do major library refactoring / improvements until most of the basics in iOS 14 is working, since I don't think this is a wide spread issue.

**Comment by Popolechien** [comment]
Interesting. I'm happy to give people that kind of answer when they reach out, but if the source file is ok and gets corrupted on the way, is it reasonable that I suggest they try and download again or is there some bigger issue (e.g., they should not interrupt the download or simply switch to a nopic/mini version to be on the safe side)?

**Comment by kelson42** [comment]
> In general, we should close issues without detail as unable to reproduce after a certain amount of time, since we really don't know what is wrong.

I agree, if there is nothing we can do and there is no upstream bug, then it makes no sense to keep a ticket open. But we should do our best to avoid this kind of situation by allowing the users to share with us the details necessary for us to understand the problem.

> This one, however, I think I have a pretty good idea about what's going on. Right now, the app does not handle the file corrupted case very well. This is why you see the file downloaded to 100% (or like 99.9%), then the progress hangs / stalls. The app moves the file from downloading to on device upon successfully opening a file. And it won't be moved if the file cannot be opened.

Why we don't move the file and then open it? Sounds a better separation of duty and would allow to make the diference between a corrupted ZIM file and a download problem?!

> What would be nice is to show user the file name and let them know you got this file, but we cannot open it. And offer to checksum if possible, so that we know if it is the download or the file itself that is problematic.

If Kiwix can not open a ZIM file, we should have a dedicated handling. I would recommend to open a ticket to discuss this.

> Right now, I am not gonna do major library refactoring / improvements until most of the basics in iOS 14 is working, since I don't think this is a wide spread issue.

I'm not sure if we share the same definition of "wide spreading", but AFAIK people keep complaining about that and the ticket is pinned for a good reason. This is definitly from my perspective part of the TOP3 bugs in Kiwix iOS.

**Comment by automactic** [comment]
@kelson42:

> Why we don't move the file and then open it? Sounds a better separation of duty and would allow to make the diference between a corrupted ZIM file and a download problem?!

We are already moving the file and attempt to open it. What could be improved is what happens if the file failed to open

> If Kiwix can not open a ZIM file, we should have a dedicated handling. I would recommend to open a ticket to discuss this.

Like I already said, there could be improvements. But not until the next round of library focused improvements

**Comment by automactic** [comment]
@Popolechien & @kelson42:

How often is download related issue really being reported? I have no visibility of how many user is having what kind of feedbacks. From my perspective, there is this current issue right here and a few of others that reported issues that might be related to this.

There could be so many reason why download failed or a downloaded file cannot be opened. Apart from what is suggested above, perhaps use a computer to download file and copy to device could also be a good idea.

Also, on the technical level, the download task is handed over to the system daemon (urlsessiond) since the app will be paused when it goes into background. This way user can use the phone as usual and don't have to keep the app foreground. I personally start a download before bed time and leave the phone plugged in over night. The down side is the OS have more control to the downloading process. For example, the system might decide low power mode is on and it will be better to download it later when plugged in.

**Comment by Popolechien** [comment]
OK thanks @automactic for the run down. At the moment I'd say we get 1-2 emails a day on that very issue, so it's fairly significant.

**Comment by automactic** [comment]
@Popolechien what does these email all say? What are some of the common symptoms?

**Comment by Popolechien** [comment]
@automactic They vary in the amount of details but re-reading the last few I stand corrected - they were more related to search and a couple of people wrote back to say it had now been fixed. Regarding download I'll send you a couple of comments via slack next time we get them but honestly most people stop at "it does not work. plz advise"

**Comment by automactic** [comment]
@Popolechien Any additional feedback? If not, we perhaps should close this, because there aren't anything I could do to further pinpoint the issue or figure out what's going on

**Comment by Popolechien** [comment]
Nothing for the past two weeks. Let's close this.
